src/transactions/test/TxResultsTests.cpp

```diff
@@ -113,6 +112,236 @@ TEST_CASE("txresults", "[tx][txresults]")
+    auto nonEnoughBalance =
+        root.create("nonEnoughBalance", lm.getLastMinBalance(0));
+    auto enoughBalance = root.create("enoughBalance", startAmount);
+    auto extraSignature = TestAccount{*app, getAccount("extraSignature")};
+
+    auto validateError = [&](TestAccount const& account,
+                             Transaction const& rawTx, int authLevel,
+                             ValidationResult const& validationResult,
+                             TransactionResult const& applyResult = {}) {
+        auto tx = account.unsignedTx(rawTx);
+
+        for_all_versions(*app, [&] {
+            auto version = [&]() {
+                LedgerTxn ltx(app->getLedgerTxnRoot());
+                return ltx.loadHeader().current().ledgerVersion;
+            }();
+
+            SECTION("no signatures")
+            {
+                if (version == 7 || authLevel < 1)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH});
+                }
+            }
+
+            SECTION("one signature")
+            {
+                account.sign(tx);
+                validateTxResults(tx, *app, validationResult, applyResult);
+            }
+
+            SECTION("double signature")
+            {
+                account.sign(tx);
+                account.sign(tx);
+
+                if (version == 7 || authLevel < 2)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH_EXTRA});
+                }
+            }
+
+            SECTION("extra signature")
+            {
+                account.sign(tx);
+                extraSignature.sign(tx);
+
+                if (version == 7 || authLevel < 2)
+                {
+                    validateTxResults(tx, *app, validationResult, applyResult);
+                }
+                else
+                {
+                    validateTxResults(tx, *app,
+                                      {validationResult.fee, txBAD_AUTH_EXTRA});
+                }
+            }
+        });
+    };
+
+    SECTION("transaction errors")
+    {
+        SECTION("non existing")
+        {
+            SECTION("all possible bugs")
+            {
+                auto rawTx = nonExisting.rawTx({}, 4);
+                rawTx.timeBounds.activate().minTime = getCloseTime() + 1;
+                rawTx.timeBounds.activate().maxTime = getCloseTime() - 1;
+                rawTx.fee = 1;
+
+                validateError(nonExisting, rawTx, 0, {1, txMISSING_OPERATION});
+            }
+
+            SECTION("txMISSING_OPERATION solved")
```

Comment:
the section names are super confusing: "XYZ solved" doesn't say what the section is testing, this one for example should be "txTOO_EARLY"